File: causeway/sdks/go/context.go
package raceway

import (
	"context"
	"fmt"
	"os"
	"sync"
	"sync/atomic"
)

var goroutineCounter uint64

// RacewayContext holds trace-related state for the current execution
type RacewayContext struct {
	TraceID     string  // UUID identifying this trace
	GoroutineID string  // Unique ID for this goroutine/execution chain
	ParentID    *string // ID of parent event (nil for roots)
	RootID      *string // ID of root event in this chain
	Clock       uint64  // Logical clock for this event chain
	mu          sync.Mutex
}

// Context key for storing RacewayContext
type contextKey string

const racewayContextKey contextKey = "raceway_context"

// NewRacewayContext creates a new context for a trace
func NewRacewayContext(traceID string) *RacewayContext {
	// Generate unique goroutine ID
	gid := atomic.AddUint64(&goroutineCounter, 1)
	goroutineID := fmt.Sprintf("go-%d-%d", os.Getpid(), gid)

	return &RacewayContext{
		TraceID:     traceID,
		GoroutineID: goroutineID,
		ParentID:    nil,
		RootID:      nil,
		Clock:       0,
	}
}

// Update updates the context after an event (thread-safe)
func (rc *RacewayContext) Update(eventID string, isFirstEvent bool) {
	rc.mu.Lock()
	defer rc.mu.Unlock()

	if isFirstEvent && rc.RootID == nil {
		rc.RootID = &eventID
	}
	rc.ParentID = &eventID
	rc.Clock++
}

// GetRacewayContext extracts RacewayContext from context.Context
func GetRacewayContext(ctx context.Context) *RacewayContext {
	if ctx == nil {
		return nil
	}
	raceCtx, ok := ctx.Value(racewayContextKey).(*RacewayContext)
	if !ok {
		return nil
	}
	return raceCtx
}

// WithRacewayContext adds RacewayContext to a context.Context
func WithRacewayContext(ctx context.Context, raceCtx *RacewayContext) context.Context {
	return context.WithValue(ctx, racewayContextKey, raceCtx)
}

File: sdks/python/raceway/context.py
"""Context management for Raceway SDK using contextvars."""

import os
import uuid
from contextvars import ContextVar
from dataclasses import dataclass, field
from typing import Optional, List, Tuple


@dataclass
class RacewayContext:
    """Context for tracking trace state across async/sync execution."""

    trace_id: str
    execution_id: str  # Unique ID for this execution chain (like goroutine ID)
    parent_id: Optional[str] = None
    root_id: Optional[str] = None
    clock: int = 0
    span_id: str = field(default_factory=lambda: uuid.uuid4().hex[:16])
    parent_span_id: Optional[str] = None
    distributed: bool = False
    clock_vector: List[Tuple[str, int]] = field(default_factory=list)
    tracestate: Optional[str] = None


# Context variable for automatic propagation (works with threading and asyncio)
_raceway_context: ContextVar[Optional[RacewayContext]] = ContextVar(
    'raceway_context',
    default=None
)


def create_context(
    trace_id: Optional[str] = None,
    *,
    span_id: Optional[str] = None,
    parent_span_id: Optional[str] = None,
    distributed: bool = False,
    clock_vector: Optional[List[Tuple[str, int]]] = None,
    tracestate: Optional[str] = None,
) -> RacewayContext:
    """
    Create a new Raceway context with unique execution ID.

    Args:
        trace_id: Trace ID to use (generates new UUID if not provided)

    Returns:
        New RacewayContext instance
    """
    if trace_id is None:
        trace_id = str(uuid.uuid4())

    # Generate unique execution ID using UUID (matching Node SDK approach)
    # Format: python-<pid>-<uuid-first-8-chars>
    execution_id = f"python-{os.getpid()}-{str(uuid.uuid4())[:8]}"

    return RacewayContext(
        trace_id=trace_id,
        execution_id=execution_id,
        parent_id=None,
        root_id=None,
        clock=0,
        span_id=span_id or uuid.uuid4().hex[:16],
        parent_span_id=parent_span_id,
        distributed=distributed,
        clock_vector=clock_vector.copy() if clock_vector else [],
        tracestate=tracestate,
    )


def set_context(ctx: RacewayContext) -> None:
    """Set the current Raceway context."""
    _raceway_context.set(ctx)


def get_context() -> Optional[RacewayContext]:
    """Get the current Raceway context."""
    return _raceway_context.get()


def update_context(event_id: str, is_first_event: bool) -> None:
    """
    Update the current context after capturing an event.

    Args:
        event_id: ID of the captured event
        is_first_event: Whether this is the first event in the execution chain
    """
    ctx = get_context()
    if ctx is None:
        return

    # Set root ID if this is the first event
    if is_first_event and ctx.root_id is None:
        ctx.root_id = event_id

    # Update parent ID and increment clock
    ctx.parent_id = event_id
    ctx.clock += 1
